src/Symfony/Component/Message/HandlerLocator.php

```diff
@@ -19,15 +20,19 @@
  */
 class HandlerLocator implements HandlerLocatorInterface
 {
+    /** @var ContainerInterface */
+    private $serviceLocator;
+
     /**
      * Maps a message (its class) to a given handler.
      *
      * @var array
      */
     private $messageToHandlerMapping;
 
-    public function __construct(array $messageToHandlerMapping = array())
+    public function __construct(ContainerInterface $serviceLocator, array $messageToHandlerMapping = array())
```

Comment:
Cant the locator do the mapping as well? `$handler = $locator->get(get_class($message))`

Meaning the multiple handlers per message are normalized to a chain/collection handler at compile time.